src/protocol/modulus_conversion/gen_random.rs

```diff
@@ -0,0 +1,166 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::{
+        mesh::{Gateway, Mesh},
+        prss::SpaceIndex,
+        Identity,
+    },
+    protocol::{context::ProtocolContext, RecordId, Step},
+    secret_sharing::Replicated,
+};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, Default)]
+pub struct ReplicatedBinary(bool, bool);
+
+impl ReplicatedBinary {
+    #[must_use]
+    #[allow(dead_code)]
+    pub fn new(a: bool, b: bool) -> Self {
+        Self(a, b)
+    }
+}
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub enum ModulusConversionStep {
+    Xor1,
+    Xor2,
+}
+
+#[derive(Debug)]
+pub struct GenRandom {
+    input: ReplicatedBinary,
+}
+
+impl GenRandom {
+    #[allow(dead_code)]
+    pub fn new(input: ReplicatedBinary) -> Self {
+        Self { input }
+    }
+
+    fn local_secret_share<F: Field>(
+        input: ReplicatedBinary,
+        channel_identity: Identity,
+    ) -> (Replicated<F>, Replicated<F>, Replicated<F>) {
+        match channel_identity {
+            Identity::H1 => (
+                Replicated::new(F::from(u128::from(input.0)), F::ZERO),
+                Replicated::new(F::ZERO, F::from(u128::from(input.1))),
+                Replicated::new(F::ZERO, F::ZERO),
+            ),
+            Identity::H2 => (
+                Replicated::new(F::ZERO, F::ZERO),
+                Replicated::new(F::from(u128::from(input.0)), F::ZERO),
+                Replicated::new(F::ZERO, F::from(u128::from(input.1))),
+            ),
+            Identity::H3 => (
+                Replicated::new(F::ZERO, F::from(u128::from(input.1))),
+                Replicated::new(F::ZERO, F::ZERO),
+                Replicated::new(F::from(u128::from(input.0)), F::ZERO),
+            ),
+        }
+    }
+
+    async fn xor<F: Field, M: Mesh, G: Gateway<M, S>, S: Step + SpaceIndex>(
+        a: Replicated<F>,
+        b: Replicated<F>,
+        ctx: &ProtocolContext<'_, G, S>,
+        step: S,
+        record_id: RecordId,
+    ) -> Result<Replicated<F>, BoxError> {
+        let result = ctx.multiply(record_id, step).await.execute(a, b).await?;
+
+        Ok(a + b - (result * F::from(2)))
+    }
+
+    #[allow(dead_code)]
+    pub async fn execute<F: Field, M: Mesh, G: Gateway<M, S>, S: Step + SpaceIndex>(
+        &self,
+        ctx: &ProtocolContext<'_, G, S>,
+        record_id: RecordId,
+        step1: S,
```

Comment:
this doesn't look right and ergonomic, I need to think and talk to Martin to see if there is a better way to do it

Reply:
Thank you! Yes, this feels strange... passing in 2 steps... something is off somehow.

Reply:
One idea I had here is to pass in just the first step (and provide an interface that allows someone to construct the first step without having to know about Xor1 vs. Xor2).  Then provide some means of generating the Xor2 step from that.

I don't know how to manage that, to be honest, I'm still pondering it.